theories/algebra/IntDiv.ec

```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
 by rewrite {1}(divz_eq (i %% k^n) (k^p)) modz_dvd_pow.
 qed.
 
-lemma nosmt modz_pow2_div n p i: 0 <= i => 0 <= p <= n =>
-  (i %% 2^n) %/ 2^p = (i %/ 2^p) %% 2^(n-p).
-proof. admitted.
+lemma nosmt divzMr i a b :  
+  0 <= a => 0 <= b => i %/ (a * b) = i %/ a %/ b.
+proof.
+move=> ge0_a ge0_b.
+case (a = 0); first by smt(div0z). 
```

Comment:
I just can't see how `smt(div0z)` is solving this. Are we missing core lemmas about `%%` that are true in the `SMTLIB` model? (Don't answer that, simply tell me if you have a non-smt proof.)

Reply:
This one truly has me worried: Alt-Ergo has messed up Euclidean division in the past and we should understand what the SMT solvers are doing.

Reply:
So this is `by rewrite divz0 div0z.`

The fact that `smt(divz0)` finds this proof means that one of the following is true:
1) SMTLIB specifies division by 0; or
2) `divz0` is a trivial (for SMT) consequence of the definition of `edivz`.

If neither is true, then we should dig a bit deeper.

Reply:
Yes, I also noticed this oddity (however, instead of being critical like you, I simply assumed it should be alright) as I first did this with regular rewriting. However, to shorten proofs, I often replace annoying rewrites with only relatively simple or rudimentary (e.g. associativity, commutativity, multiplication by 0, etc.) by an SMT call with the appropriate lemmas given. I find the 'appropriate' lemmas by simply passing all the lemmas I used in the rewrite (that are slightly non-trivial) and then trimming them down to see what smt minimally needs. This led to the `smt(div0z)` in this case, even though, as you can see in the current code (and as you already noted), the rewrite indeed also requires `divz0`. Nevertheless, for now, I removed all SMT calls just to be sure.

Reply:
For ring/field equalities, you can use the `ring` / `field` tactic.